### ISSUE TITLE
ConcurrentBag causes memory leak

### ISSUE THREAD

**Issue by kopfsick** [AutoMapper/AutoMapper#1691]
After recently upgrading from 4.2 to 5.1 our (FitNesse) test suite started running out of memory. In this testsuite the bootstrapper, and therefore our Automapper config, of our application is executed several (read thousands of) times.
I've tracked the problem to TypeMap's ConcurrentBag<PropertyMap>. The ConcurrentBag class has some issues with memory leaks (http://stackoverflow.com/questions/10850443/is-there-a-memory-leak-in-the-concurrentbagt-implementation).
In short, when initializing Automapper (both using static and instance api) multiple times on the same thread, all typemaps and propertymaps will be kept alive in a ThreadLocal until the thread is killed.

Anybody else experience this and/or have a proper solution?
Are there any way to properly dispose and clear the Automapper configuration?

I've temporarily fixed it by using the static API and only initializing Automapper if I detect that it is not initialized.

**Comment by lbargaoanu** [comment]
TypeMap doesn't need to be thread safe. That's old code. PR?

**Comment by kopfsick** [comment]
Tested after removing the use of ConcurrentBag. It improved things a bit, but there's also a memory issue with 
`private static readonly ConcurrentDictionary<TypeDetails, Dictionary<MemberInfo, IEnumerable<SourceToDestinationMapperAttribute>>> Cache = new ConcurrentDictionary<TypeDetails, Dictionary<MemberInfo, IEnumerable<SourceToDestinationMapperAttribute>>>();`

in `SourceToDestinationNameMapperAttributesMember`. A static cache is a bad idea imho.

**Comment by kopfsick** [comment]
I'm not too well known with the Automapper code, but it looks as though that dictionary could be just private non-static without to much issues....

**Comment by lbargaoanu** [comment]
There is no need for it to be static.

**Comment by kopfsick** [comment]
I'll refactor and test more tomorrow and make a PR when things look OK.

**Comment by kopfsick** [comment]
#1692 
Thanks for your guidance.

**Comment by mbeauchamp7** [comment]
When do you think this will be released? We are having major memory leaks! Thanks

**Comment by lbargaoanu** [comment]
@mbeauchamp7 The problem is most likely in your app. This is a particular case, for testing. There is no memory leak that we know of in normal usage. But if you have a repro, we want to see it.

**Comment by mbeauchamp7** [comment]
@lbargaoanu We're going to look deeper into our solution if the problem is actually in our hands. We have a pretty complex automapper configuration so this might take a while... I'm not really sure what to show you exactly but I have a dump file that gave us a clue that this might be coming from the automapper.

**Comment by lbargaoanu** [comment]
That might help. Hopefully you're on the latest version, because that's the only one actively developed.

**Comment by mbeauchamp7** [comment]
We're actually using 5.1.1, to who can I send the dump file?

**Comment by lbargaoanu** [comment]
You'll have to upload it somewhere and post a link.

**Comment by mbeauchamp7** [comment]
[dump_oct4_16.txt](https://github.com/AutoMapper/AutoMapper/files/511791/dump_oct4_16.txt)

If you look at the bottom, you can see what uses most of the memory.
Hope this helps!

**Comment by lbargaoanu** [comment]
Maybe. But the full dump would be more helpful. Or a way to replicate your results.

**Comment by lbargaoanu** [comment]
I don't see it. But I don't understand what code results in that heap. 24720 type maps? :)

**Comment by mbeauchamp7** [comment]
Yeah it's some serious leak lol... Anyways thanks for your time! I'll look into our code to fix this and let you know what it was. (bit busy with personal stuff atm)
